### PR TITLE
DAOS-4170 control: Simplify pool creation

### PR DESCRIPTION
**Pull request by mjmac**
The `dmg pool create` command now has a --size parameter which
can be used to specify the total size of a pool. The --scm-ratio
and --num-ranks parameters may be optionally specified if the
defaults (6% SCM:NVMe, all ranks) are not suitable.

The existing --scm-size and --nvme-size flags remain, but should
generally not be used and may be removed eventually.

The pool create output has also been updated to provide more
detailed information about how the pool was created.

The default value for --nsvc is now 3, in order to create resilient
pools by default.

### CODE REVIEW

File: src/tests/ftest/util/dmg_utils.py
Comment (ajskaar) [inline src/tests/ftest/util/dmg_utils.py:88]: Perhaps we can use a single _get_result() function that checks if the json output flag was set or not. This if statement was removed from down below and we want to be able to use this test utility to output in json or plain text. The other modules in this file will be updated to make use of this function also -- likely in DAOS-4908, which I have put on hold until this ticket lands. @mjmac  @phender

File: src/tests/ftest/util/dmg_utils.py
Comment (ajskaar) [inline src/tests/ftest/util/dmg_utils.py:413]: Do we want to always return json format? Maybe we keep the if condition? I'd prefer the output from these test utilities to look the exact same as if running dmg directly on cmd line. I plan to use these modules to verify dmg json output and non json output are as expected (and the same). @mjmac  @phender

Comment (mjmac) [reply src/tests/ftest/util/dmg_utils.py:413]: I would really prefer to not perpetuate these dependencies on stable plaintext output in the python test harness. We already have unit/integration tests in the control plane for verifying that dmg's pretty printers work as expected for various outputs, and it's much easier to keep those tests in sync with any human-consumable output updates. In my opinion, it would be a better use of test engineering time to work on migrating the harness to consume JSON in order to reduce test fragility and maintenance overhead.

**Comment by mjmac** [comment]
Example output:
```
[macdonal@wolf-61 daos]$ dmg pool create --size=500GB
Creating DAOS pool with automatic storage allocation: 500 GB NVMe + 6.00% SCM
Pool created with 6.00% SCM/NVMe ratio
---------------------------------------
  UUID          : 4f563c9d-438c-43b7-b29f-a6070cb0d5b7
  Service Ranks : [1-3]
  Storage Ranks : [0-15]
  Total Size    : 530 GB
  SCM           : 30 GB (1.9 GB / rank)
  NVMe          : 500 GB (31 GB / rank)
```

File: src/tests/ftest/util/dmg_utils.py
Comment (phender) [reply src/tests/ftest/util/dmg_utils.py:88]: My vote would be for keeping the `_get_json_result()` - with the addition of returning `json.loads(self.result.stdout)` - and eventually using it to replace the `_get_result()` method calls in all the dmg command helper methods.  By not overriding the `_get_result()` method we leave the option open to still get non-json output if we ever need it.

Comment (mjmac) [reply src/tests/ftest/util/dmg_utils.py:88]: Yep. Forgot to make this change when I pushed the latest update. I'll push it after dealing with any other review feedback.

File: src/control/cmd/dmg/pool.go
Comment (kjacque) [inline src/control/cmd/dmg/pool.go:119]: Just a suggestion, but for improved clarity I think it'd be nice to break out the blocks into helper functions (updateReqStorageAuto, updateReqStorageManual, or something like that).

File: src/control/cmd/dmg/pool_test.go
Comment (kjacque) [inline src/control/cmd/dmg/pool_test.go:123]: Should probably have cases for all the bad combinations, right? i.e. A version of this with --scm-size

File: src/control/server/mgmt_pool_test.go
Comment (kjacque) [inline src/control/server/mgmt_pool_test.go:144]: nit - spelling
```suggestion
	scmTooSmallReq := uint64(float64(scmTooSmallTotal) * scmTooSmallRatio)
```

File: src/control/system/database_pools.go
Comment (kjacque) [inline src/control/system/database_pools.go:55]: The opposite of what I'd expect, as far as public/private. I'm curious why the strings are saved at all. Can't we generate them from the RankSets?

**Review by kjacque — DISMISSED** [review]
This looks good. I have some minor comments but nothing I'd block on.

File: src/control/server/mgmt_pool.go
Comment (tanabarr) [inline src/control/server/mgmt_pool.go:204]: shouldn't "the requested number of" be "request number (or all if not)"?

File: src/control/system/database_pools.go
Comment (tanabarr) [reply src/control/system/database_pools.go:55]: the canonical form of a rank set is its ranged string so probably isn't necessary to store the rich type IMO

File: src/control/system/database_pools.go
Comment (tanabarr) [inline src/control/system/database_pools.go:143]: will be useful for system storage query

File: src/proto/mgmt/pool.proto
Comment (tanabarr) [inline src/proto/mgmt/pool.proto:48]: would be cleaner to have oneof config config {auto, manual} with the relevant sub-messages defined

**Review by tanabarr — APPROVED** [review]

File: src/control/system/database_pools.go
Comment (mjmac) [reply src/control/system/database_pools.go:55]: It's more space efficient to store the string form. Also much easier to serialize for snapshots. The public/private here isn't really for external users, it's to control which fields are exposed to json. I guess I could use tags to make this more explicit.

File: src/control/cmd/dmg/pool.go
Comment (mjmac) [reply src/control/cmd/dmg/pool.go:119]: Agreed that this could be cleaner. Longer-term, I am thinking that the manual options may go away though. I don't really see valid use cases for them other than compatibility with older scripts.

File: src/control/server/mgmt_pool_test.go
Comment (mjmac) [reply src/control/server/mgmt_pool_test.go:144]: Gah, fixed that in one repo and forgot to bring it over to the one I pushed from. Fixed.

File: src/control/cmd/dmg/pool_test.go
Comment (mjmac) [reply src/control/cmd/dmg/pool_test.go:123]: Done.

File: src/proto/mgmt/pool.proto
Comment (mjmac) [reply src/proto/mgmt/pool.proto:48]: The thing is, the scm/nvme fields are always filled in when sending the request down via dRPC, so it's not really an either/or situation.

File: src/control/system/database_pools.go
Comment (mjmac) [reply src/control/system/database_pools.go:143]: Yep, that's what I was thinking. I figured I'd add this now because I was already in there, and having this in place will give us something to build on.

File: src/control/server/mgmt_pool.go
Comment (mjmac) [reply src/control/server/mgmt_pool.go:204]: Updated comment.

File: src/tests/ftest/util/pool_test_base.py
Comment (ajskaar) [inline src/tests/ftest/util/pool_test_base.py:101]: For whatever reason we have been using "svcn" (n in back)as the name of the our pool attributes, as opposed to the binary dmg cmd which was using "nsvc" (n in front). The pool create tests failed last time around for this reason, and probably will again until this is fixed. Option 1: You will need to do "pool.svcn.update(nsvc), or option 2 (probably better): to change the local variable name for and description also (lines 74 and and 88), then call "pool,svcn.update(svcn)" ---- OR option 3 (my personal favorite): we update the name of this property in the TestPool class to something that matches the new field name used by binary (I think you picked NumSvcResps or something) . The last option would require several changes to other tests/yaml files, Option 2 likely the easiest for now, but option 3 seems best...our naming conventions are all over the place.... @mjmac @phender

Comment (mjmac) [reply src/tests/ftest/util/pool_test_base.py:101]: Oops, good catch. I usually run pool/simple_create_delete.py after making changes to sanity-check, but I guess that one doesn't involve this code.

File: src/control/server/faults.go
Comment (tanabarr) [inline src/control/server/faults.go:68]: To me this isn't clearly talking about the number of service ranks, it might be made more clear to the user by rewording to "number of service ranks" otherwise it might be confused with service replica rank value. Changing the naming to "service ranks" rather than service replicas it also in line with other changes to the naming elsewhere. Renaming the fault name to "FaultPoolInvalidNumServiceRanks" would also improve clarity.

**Review by tanabarr — APPROVED** [review]

File: src/control/system/database_pools.go
Comment (tanabarr) [inline src/control/system/database_pools.go:62]: Raft term field was requested in DAOS-6088, presumably that should be added here as well so that pool service replica ranks update events can provide raft term value.

**Review by tanabarr — DISMISSED** [review]

File: src/control/system/database_pools.go
Comment (mjmac) [reply src/control/system/database_pools.go:62]: Hmm, yes, I think you're right about that. We want it to be sent as part of the message so that we can ensure that we're only making an update if the term is greater than the current value. That having been said, I'd prefer to add that in another PR that is focused on handling the service rank events.

File: src/control/server/faults.go
Comment (mjmac) [reply src/control/server/faults.go:68]: We definitely have some inconsistency throughout the code. I've been moving toward "service replicas" in general rather than "service ranks", as it more clearly conveys their role IMO.

For what it's worth, the dmg help for pool create has referred to "service replicas" for a long time now: https://github.com/daos-stack/daos/blob/master/src/control/cmd/dmg/pool.go#L81

**Review by kjacque — DISMISSED** [review]

**Review by kjacque — DISMISSED** [review]

**Review by tanabarr — APPROVED** [review]

**Review by kjacque — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]

**Comment by mjmac** [comment]
One known failure due to [DAOS-6442](https://jira.hpdd.intel.com/browse/DAOS-6442). That test is being disabled in master as it is causing failures on most PRs now.

**Pull request merged**
